# Data Quality Preparation
In this section I will be cleaning and preparing the vehicles and trip tables for the prediction model. It will be divided into two parts as they will be cleaned and prepared seperately.

#### Introduction:

### Trip Dataset
Each row represents one trip(route)
#### Understanding the features:
* DATASOURCE: Unique Bus Operator Code
* DAYOFSERVICE: Day of service. One day of service could last more than 24 hours
* TRIPID: Unique Trip code
* LINEID: Unique Line code
* ROUTEID: Unique route code
* DIRECTION: Route direction: (2)IB = inbound / going / northbound / eastbound, (1)OB = outbound / back / southbound / westbound
* PLANNEDTIME_ARR: Planned arrival time of the trip, in seconds
* PLANNEDTIME_DEP: Planned departure time of the trip, in seconds
* ACTUALTIME_ARR: Actual arrival time of the trip, in seconds
* ACTUALTIME_DEP: Actual departure time of the trip, in seconds
* BASIN: basin code
* TENDERLOT: tender lot
* SUPPRESSED: The whole trip has been supressed (0 = achieved, 1 = suppressed)
* JUSTIFICATIONID: Fault code
* LASTUPDATE: Time of the last record update 
* NOTE: Free note

### Vehicles Dataset
Each record of this table represents the service of one vehicle for one day of service and tells about
the overall distance and time worked by the associated vehicle in that specific day.

#### Understanding the features:
* DATASOURCE: Unique Bus Operator Code
* DAYOFSERVICE: Day of service
* VEHICLEID: Unique vehicle code arriving at this stop point
* DISTANCE: Distance travelled by the vehicle in the corresponding day
* MINUTES: Time worked by the veihcle in the corresponding day
* LASTUPDATE: Time of the last record update
* NOTE: Free note

In [1]:
# Importing the modules that might be used
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as patches
import dask.dataframe as dd
import sqlite3
from sqlalchemy import create_engine
from pprint import pprint

# ignore warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
leave_times = pd.read_feather('/home/faye/data/leavetimes.feather')

In [3]:
# Loading in the data
trips = pd.read_csv('/home/faye/data/rt_trips_DB_2018.txt', sep=';',error_bad_lines=False)
vehicles = pd.read_csv('/home/faye/data/rt_vehicles_DB_2018.txt', sep=';',error_bad_lines=False)
# leave_times_db = create_engine('sqlite:///rt_leavetimes_DB_2018.db')
# con = sqlite3.connect('/home/faye/notebooks/rt_leavetimes_DB_2018.db')

leavetimes = pd.read_sql("select * from chunk_sql group by TRIPID order by max(PROGRNUMBER);", con)

## Trips 
*Exploring the data*
<br><br>

In [4]:
trips.head(50)

DATASOURCE        DAYOFSERVICE   TRIPID LINEID  ROUTEID  DIRECTION  \
0          DB  07-FEB-18 00:00:00  6253783     68    68_80          1   
1          DB  07-FEB-18 00:00:00  6262138    25B  25B_271          2   
2          DB  07-FEB-18 00:00:00  6254942    45A   45A_70          2   
3          DB  07-FEB-18 00:00:00  6259460    25A  25A_273          1   
4          DB  07-FEB-18 00:00:00  6253175     14    14_15          1   
5          DB  07-FEB-18 00:00:00  6248240    77A   77A_28          2   
6          DB  07-FEB-18 00:00:00  6251760     39    39_21          2   
7          DB  07-FEB-18 00:00:00  6262909     16    16_20          1   
8          DB  07-FEB-18 00:00:00  6251147    40D  40D_102          1   
9          DB  07-FEB-18 00:00:00  6249435    27B   27B_34          2   
10         DB  07-FEB-18 00:00:00  6251483    142    142_9          2   
11         DB  07-FEB-18 00:00:00  6254310     83    83_22          2   
12         DB  07-FEB-18 00:00:00  6258441     14    14_15          1   
13         DB  07-FEB-18 00:00:00  6247622    130   130_10          1   
14         DB  07-FEB-18 00:00:00  6256532     15    15_16          1   
15         DB  07-FEB-18 00:00:00  6253924    46A   46A_74          1   
16         DB  07-FEB-18 00:00:00  6256349     33    33_44          1   
17         DB  07-FEB-18 00:00:00  6259785      7     7_70          2   
18         DB  07-FEB-18 00:00:00  6246028    39A   39A_40          1   
19         DB  07-FEB-18 00:00:00  6261203    39A   39A_43          2   
20         DB  07-FEB-18 00:00:00  6262164    46A   46A_74          1   
21         DB  07-FEB-18 00:00:00  6258154     49    49_28          2   
22         DB  07-FEB-18 00:00:00  6253554      1     1_40          2   
23         DB  07-FEB-18 00:00:00  6246745     15    15_16          1   
24         DB  07-FEB-18 00:00:00  6246014    123   123_34          1   
25         DB  07-FEB-18 00:00:00  6254029     83    83_22          2   
26         DB  07-FEB-18 00:00:00  6256059     41     41_7          2   
27         DB  07-FEB-18 00:00:00  6257123    67X   67X_38          1   
28         DB  07-FEB-18 00:00:00  6259641     59    59_10          1   
29         DB  07-FEB-18 00:00:00  6245836      9      9_5          1   
30         DB  07-FEB-18 00:00:00  6252377    130   130_10          1   
31         DB  07-FEB-18 00:00:00  6254347     40    40_27          1   
32         DB  07-FEB-18 00:00:00  6251013    39A   39A_40          1   
33         DB  07-FEB-18 00:00:00  6260186    239   239_28          2   
34         DB  07-FEB-18 00:00:00  6253738     76    76_37          2   
35         DB  07-FEB-18 00:00:00  6247623    130   130_11          2   
36         DB  07-FEB-18 00:00:00  6260764     84    84_27          1   
37         DB  07-FEB-18 00:00:00  6261193     53    53_20          1   
38         DB  18-FEB-18 00:00:00  6268949    185   185_55          1   
39         DB  18-FEB-18 00:00:00  6268951    185   185_55          1   
40         DB  18-FEB-18 00:00:00  6268953    185   185_55          1   
41         DB  18-FEB-18 00:00:00  6268387     16    16_20          1   
42         DB  18-FEB-18 00:00:00  6277453     16    16_20          1   
43         DB  18-FEB-18 00:00:00  6277810     16    16_20          1   
44         DB  18-FEB-18 00:00:00  6272110     41     41_3          1   
45         DB  18-FEB-18 00:00:00  6278258     41     41_3          1   
46         DB  18-FEB-18 00:00:00  6270605     41     41_3          1   
47         DB  18-FEB-18 00:00:00  6279865     41     41_3          1   
48         DB  18-FEB-18 00:00:00  6275294     41     41_3          1   
49         DB  18-FEB-18 00:00:00  6274325     41     41_3          1   

    PLANNEDTIME_ARR  PLANNEDTIME_DEP  ACTUALTIME_ARR  ACTUALTIME_DEP   BASIN  \
0             87245            84600         87524.0         84600.0  BasDef   
1             30517            26460         32752.0             NaN  BasDef   
2             35512            32100   

In [5]:
trips.tail(50)

DATASOURCE        DAYOFSERVICE   TRIPID LINEID  ROUTEID  DIRECTION  \
2182587         DB  14-MAY-18 00:00:00  6759988     13    13_60          1   
2182588         DB  14-MAY-18 00:00:00  6760080    140   140_19          1   
2182589         DB  14-MAY-18 00:00:00  6760086    111    111_8          2   
2182590         DB  14-MAY-18 00:00:00  6760100    123   123_34          1   
2182591         DB  14-MAY-18 00:00:00  6760154     76    76_37          2   
2182592         DB  14-MAY-18 00:00:00  6763776     33    33_72          2   
2182593         DB  14-MAY-18 00:00:00  6763777     33    33_45          1   
2182594         DB  14-MAY-18 00:00:00  6763797     84    84_31          2   
2182595         DB  14-MAY-18 00:00:00  6763806    151   151_15          1   
2182596         DB  14-MAY-18 00:00:00  6763831    151   151_16          1   
2182597         DB  14-MAY-18 00:00:00  6763881    17A   17A_17          2   
2182598         DB  14-MAY-18 00:00:00  6757370     67     67_6          2   
2182599         DB  14-MAY-18 00:00:00  6757400     47   47_139          1   
2182600         DB  14-MAY-18 00:00:00  6757403     47   47_135          2   
2182601         DB  14-MAY-18 00:00:00  6757455    161   161_51          2   
2182602         DB  14-MAY-18 00:00:00  6757468     33    33_44          1   
2182603         DB  14-MAY-18 00:00:00  6757509    27A    27A_4          1   
2182604         DB  14-MAY-18 00:00:00  6757514    27A    27A_5          2   
2182605         DB  14-MAY-18 00:00:00  6757515    27A    27A_4          1   
2182606         DB  14-MAY-18 00:00:00  6762371    130   130_11          2   
2182607         DB  14-MAY-18 00:00:00  6762374    130   130_10          1   
2182608         DB  14-MAY-18 00:00:00  6762495    25A  25A_270          2   
2182609         DB  14-MAY-18 00:00:00  6762520    45A   45A_60          1   
2182610         DB  14-MAY-18 00:00:00  6763184     53    53_20          1   
2182611         DB  14-MAY-18 00:00:00  6762805    140   140_21          2   
2182612         DB  14-MAY-18 00:00:00  6762895     27    27_19          1   
2182613         DB  14-MAY-18 00:00:00  6762930      9      9_7          2   
2182614         DB  14-MAY-18 00:00:00  6762969     18     18_3          1   
2182615         DB  14-MAY-18 00:00:00  6764976     7A    7A_87          2   
2182616         DB  14-MAY-18 00:00:00  6765053    54A   54A_12          2   
2182617         DB  14-MAY-18 00:00:00  6765067     26    26_24          1   
2182618         DB  14-MAY-18 00:00:00  6765072    56A   56A_31          1   
2182619         DB  14-MAY-18 00:00:00  6765081    56A   56A_30          2   
2182620         DB  14-MAY-18 00:00:00  6765101    79A   79A_28          2   
2182621         DB  14-MAY-18 00:00:00  6756423    40D  40D_102          1   
2182622         DB  14-MAY-18 00:00:00  6756429    40D  40D_102          1   
2182623         DB  14-MAY-18 00:00:00  6767663    40D  40D_104          2   
2182624         DB  14-MAY-18 00:00:00  6754588     41     41_7          2   
2182625         DB  14-MAY-18 00:00:00  6757965    151   151_15          1   
2182626         DB  14-MAY-18 00:00:00  6758027    145  145_105          2   
2182627         DB  14-MAY-18 00:00:00  6765635    130   130_11          2   
2182628         DB  14-MAY-18 00:00:00  6765649    130   130_11          2   
2182629         DB  14-MAY-18 00:00:00  6765652     53    53_21          2   
2182630         DB  14-MAY-18 00:00:00  6765662     53    53_21          2   
2182631         DB  14-MAY-18 00:00:00  6765828    45A   45A_68          1   
2182632         DB  14-MAY-18 00:00:00  6765849    123   123_36          2   
2182633         DB  14-MAY-18 00:00:00  6765469     75    75_17          1   
2182634         DB  14-MAY-18 00:00:00  6765486    33D   33D_62          2   
2182635         DB  14-MAY-18 00:00:00  6764987     70    70_60          1   
2182636         DB  14-MAY-18 00:00:00  6765012     27    27_19          1   

         PLANNEDTIME_ARR  PLA

In [6]:
# Columns and rows
trips.shape

(2182637, 16)

In [7]:
trips.dtypes

DATASOURCE          object
DAYOFSERVICE        object
TRIPID               int64
LINEID              object
ROUTEID             object
DIRECTION            int64
PLANNEDTIME_ARR      int64
PLANNEDTIME_DEP      int64
ACTUALTIME_ARR     float64
ACTUALTIME_DEP     float64
BASIN               object
TENDERLOT          float64
SUPPRESSED         float64
JUSTIFICATIONID    float64
LASTUPDATE          object
NOTE                object
dtype: object

In [8]:
trips.isnull().sum()

DATASOURCE               0
DAYOFSERVICE             0
TRIPID                   0
LINEID                   0
ROUTEID                  0
DIRECTION                0
PLANNEDTIME_ARR          0
PLANNEDTIME_DEP          0
ACTUALTIME_ARR      137207
ACTUALTIME_DEP      164551
BASIN                    0
TENDERLOT          2182637
SUPPRESSED         2178304
JUSTIFICATIONID    2178307
LASTUPDATE               0
NOTE                     0
dtype: int64

In [9]:
trips.nunique()

DATASOURCE              1
DAYOFSERVICE          360
TRIPID             658964
LINEID                130
ROUTEID               588
DIRECTION               2
PLANNEDTIME_ARR     64461
PLANNEDTIME_DEP       791
ACTUALTIME_ARR      68122
ACTUALTIME_DEP      66771
BASIN                   1
TENDERLOT               0
SUPPRESSED              1
JUSTIFICATIONID      3526
LASTUPDATE            360
NOTE                46690
dtype: int64

In [10]:
# trips_columns = list(trips.columns.values)

In [11]:
# Seperating features
categorical_columns = ['DATASOURCE','TRIPID','LINEID','ROUTEID','DIRECTION','BASIN','TENDERLOT','JUSTIFICATIONID','NOTE','SUPPRESSED']
time_columns = ['PLANNEDTIME_ARR','PLANNEDTIME_DEP','ACTUALTIME_ARR','ACTUALTIME_DEP']
datetime = ['DAYOFSERVICE','LASTUPDATE']

# Converting seconds into times and into datetime series
import datetime
for row in trips['PLANNEDTIME_ARR']:
    trips['PLANNEDTIME_ARR'] = trips['PLANNEDTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['PLANNEDTIME_DEP']:
    trips['PLANNEDTIME_DEP'] = trips['PLANNEDTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_ARR']:
    trips['ACTUALTIME_ARR'] = trips['ACTUALTIME_ARR'].replace(row, str(datetime.timedelta(seconds=row)))

for row in trips['ACTUALTIME_DEP']:
    trips['ACTUALTIME_DEP'] = trips['ACTUALTIME_DEP'].replace(row, str(datetime.timedelta(seconds=row)))

In [12]:
# Converting objects into categorical types
for column in categorical_columns:
    trips[column] = trips[column].astype('category')

# Converting objects into datetime series
for column2 in datetime:
    trips[column2] = pd.to_datetime(trips[column2])

In [13]:
trips.dtypes

DATASOURCE               category
DAYOFSERVICE       datetime64[ns]
TRIPID                   category
LINEID                   category
ROUTEID                  category
DIRECTION                category
PLANNEDTIME_ARR             int64
PLANNEDTIME_DEP             int64
ACTUALTIME_ARR            float64
ACTUALTIME_DEP            float64
BASIN                    category
TENDERLOT                category
SUPPRESSED               category
JUSTIFICATIONID          category
LASTUPDATE         datetime64[ns]
NOTE                     category
dtype: object

In [ ]:
# Converting the time columns seconds into times. The days part should be ignored.
for column3 in time_columns:
    trips[column3] = trips[column3].apply(pd.to_timedelta, unit='s')
    trips[column3] = trips[column3]- pd.to_timedelta(trips[column3].dt.days, unit='d')

In [ ]:
trips.head(10)

In [ ]:
trips_sorted = trips.sort_values(by='TRIPID')
trips_sorted.head(50)

In [ ]:
trip77a = trips_sorted.loc[trips_sorted['LINEID']=='77A']
trip77a = trip77a.loc[trips_sorted['DAYOFSERVICE']=='2018-01-03']
trip77a = trip77a.sort_values(by='PLANNEDTIME_ARR')
trip77a.head(50)

## Note about this specific route
In this route, we took the route 77a from a date (January 3, 2018) and sorted it by PLANNEDTIME_ARR. It has a total of 101 rows. The first few values of PLANNEDTIME_ARR, these may be the last few trips that ran that day and it was just sorted as the top? In any case, We see that they provide data for each route for one day with all of the times. It is a matter of matching it with the leavetimes dataset then. This will be explored at the very end. 

## Duplicates
Exploring if the trips dataset has any duplicates. There shouldn't be duplicates for any reason. The trips only siginify one route.

We see there aren't any duplicates so this is a good sign for the data. 

In [ ]:
duplicates = trips[trips.duplicated(keep=False)]
print("Number of rows that are duplicates: ", duplicates.shape[0])

## Checking descriptive columns

In [ ]:
trips[datetime].describe(datetime_is_numeric=True).T

In [ ]:
trips[time_columns].describe().T

In [ ]:
trips[categorical_columns].describe().T

We see there a couple of features that have <= 1 cardinalities:
* DATASOURCE has 1 unique value.
* TRIPID has more than 600k unique values.
* LINEID has 130 unique values.
* ROUTEID has 588 unique values.
* DIRECTION has 2 unique values.
* BASIN has 1 unique values.
* TENDERLOT has no unique values meaning that it is just null.
* JUSTIFICATIONID has 4330 unique values.
* NOTE has more than 46k unique values.
* SUPPRESSED has 4333 unique values.

**RESULT :** As we can see, DATASOURCE, BASIN, TENDERLOT and SUPPRESSED has 1 or 0 unique values. This means the same value is consistent throughout the dataset. Therefore, these will be dropped.

## Logical Integrity Tests

List of integrity tests to be performed:
* Make sure LASTUPDATE dates doesn't occur before DAYSERVICE

#### TEST 1: Ensure LASTUPDATE doesn't occur before DAYOFSERVICE

In [ ]:
test1 = trips[trips['DAYOFSERVICE'] > trips['LASTUPDATE']]
print("The number of rows that have dates that occur before DAYOFSERVICE from LASTUPDATES is: ", test1.shape[0])

#### TEST 2: Ensure timetables match the specific route with PLANNEDTIME_DEP
This is to check that the timetables match with all of the routes. We take a look at PLANNEDTIME_DEP to check this. 

In [ ]:
# Need to get a list of all of the routes and their timetables, then compare it with the routes 
# and their planned departure times

# Make a dataframe with just LINEID and PLANNEDTIME_DEP and make a list of all of the routes, order them
data = {'LINEID': trips['LINEID'], 'PLANNEDTIME_DEP': trips['PLANNEDTIME_DEP']}
line_dep = pd.DataFrame(data)
list_routes = list(line_dep.LINEID.unique())
list_routes = sorted(list_routes)

In [ ]:
trips_gtfs = pd.read_csv('/home/faye/Data-Analytics-CityRoute/Dublin_Bus_GTFS/17-07-2021/stop_times.txt', sep=',',error_bad_lines=False)

In [ ]:
trips_gtfs.head(20)

### Process to get the times:
* Got all of the stop_sequence rows
* Reseted the index (did not need to because my previous plan was scraped)
* Made three lists: service, route, and direction. Service was to indicate if it's weekday,weekend (see calendar.txt for more info), route indicates which route number, direction is for inbound and outbound (I/O)
* looped through the trips_id
    * split the column value by .
    * split the 2nd value by -
    * added first value to service
    * added first value of the 2nd value split to route
    * added fourth value to direction
    
**Extra info:**
* I parsed this from the 2021 dataset. Now, some route_id won't match with the 2018 data because there are routes that are called H3 that used to be old routes. They just changed the name. I was told it was changed because the new company that owns Dublin Bus now wants to change the system again but it was halted for reasons. So for example H3 used to be 29A (I think).  

In [ ]:
# Store the stop_sequence = 1 first
trips_1 = trips_gtfs.loc[trips_gtfs['stop_sequence']==1]

In [ ]:
# Resetting the index for this new dataframe
trips_1 = trips_1.reset_index(drop=True)

In [ ]:
trips_1.head(5)

In [ ]:
# Making a list for calendar, route number, inbound/outbound
serviceid = []
routeid = []
direction = []

for value in trips_1.trip_id:
    id_ = value.split(".")
    id_2 = id_[2].split("-")
    serviceid += [id_[1]]
    routeid += [id_2[1]]
    direction += [id_[4]]

In [ ]:
trips_1['service_id'] = np.array(serviceid)
trips_1['route_id'] = np.array(routeid)
trips_1['direction'] = np.array(direction)

In [ ]:
trips_2 = trips_1
# trips_duplicate = trips_2[trips_2.duplicated(keep=False)]

In [ ]:
trips_2 = pd.DataFrame({'departure_time': trips_2['departure_time'], 'service_id': trips_2['service_id'], 
                       'route_id': trips_2['route_id'], 'stop_headsign': trips_2['stop_headsign'], 
                       'direction': trips_2['direction']})

In [ ]:
trips_2[trips_2.duplicated(keep=False)]
trips_2 = trips_2.drop_duplicates()

In [ ]:
trips_2

In [ ]:
trips_2.loc[trips_2['route_id']=='H1'].head(50)

<br><br>
I decided against going through this local integrity test. Because there are new replacements for some of the routes, we won't be able to match them side by side. After looking at this website: https://www.transportforireland.ie/h-spine/ , it's become apparent that the departure times will be different. I will, however replace H1 > 29A, H2> 32. H3 > 31/31A. They changed the timetables so it wouldn't be possible to perform this test. It also looks like H9 is a completely new route. 

I will save the new dataset for anyone who wants to use it. This will be done in the data quality plan. 
<br><br>

## Vehicles
*Exploring the data*
<br><br>

In [ ]:
vehicles.head(10)

In [ ]:
vehicles.tail(10)

In [ ]:
vehicles.shape

In [ ]:
vehicles.dtypes

In [ ]:
vehicles.isnull().sum()

In [ ]:
vehicles.nunique()

In [ ]:
# Creating columns for different datatypes
categorical_column = ['DATASOURCE', 'VEHICLEID']
time_columns = ['MINUTES']
numerical_columns = ['DISTANCE']
datetime_columns = ['DAYOFSERVICE', 'LASTUPDATE']

In [ ]:
# Converting objects into categorical types
for column in categorical_column:
    vehicles[column] = vehicles[column].astype('category')

# Converting objects into datetime series
for column2 in datetime_columns:
    vehicles[column2] = pd.to_datetime(vehicles[column2])

# Converting the time columns seconds into times. The days part should be ignored.
for column3 in time_columns:
    vehicles[column3] = vehicles[column3].apply(pd.to_timedelta, unit='s')
    vehicles[column3] = vehicles[column3]- pd.to_timedelta(vehicles[column3].dt.days, unit='d')

In [ ]:
vehicles.head(10)

In [ ]:
vehicles.dtypes

## Duplicates
Exploring if the trips dataset has any duplicates. There shouldn't be duplicates for any reason. The trips only siginify one route.

Again, there are no duplicates in this vehicles dataset.

In [ ]:
duplicates = vehicles[vehicles.duplicated(keep=False)]
print("Number of rows that are duplicates: ", duplicates.shape[0])

## Checking descriptive columns

In [ ]:
vehicles[datetime].describe(datetime_is_numeric=True).T

In [ ]:
vehicles[categorical_column].describe().T

In [ ]:
vehicles[time_columns].describe().T

In [ ]:
vehicles[numerical_columns].describe().T

We see there a couple of features that have <= 1 cardinalities:
* DATASOURCE has 1 unique value.
* NOTE has 0 unique values and we see that 100% of the feature's values is missing.

**RESULT :** These two features will be investigated further. However, they will be dropped as constant features do no good for the targeted feature. NOTE has no unique values and has 100% missing values so this will most likely be dropped.

## Logical Integrity Tests

List of integrity tests to be performed:
* Make sure LASTUPDATE dates doesn't occur before DAYSERVICE
* Ensure that the dates match with the missing dates from the trips table

**Test 1: Ensuring LASTUPDATE dates doesn't occur before DAYOFSERVICE.**

As we can see, there are no dates that occur before DAYOFSERVICE meaning it is a good indication and there are no alterations to be done.

In [ ]:
test1 = vehicles[vehicles['DAYOFSERVICE'] > vehicles['LASTUPDATE']]
print("The number of rows that have dates that occur before DAYOFSERVICE from LASTUPDATES is: ", test1.shape[0])

**Test 2: Ensure that the dates match with the missing dates from the trips table.** 

As we can see, the dates that are missing from the vehicles table match with the missing dates from the trips table. This is a good sign. Therefore, no further actions are to be done.


In [ ]:
daysofservice_vec = vehicles['DAYOFSERVICE']
daysofservice_vec = daysofservice_vec.sort_values()
pd.date_range(start = '2018-01-01', end = '2018-12-31' ).difference(daysofservice_vec)

In [ ]:
pd.date_range(start = '2018-01-01', end = '2018-12-31' ).difference(daysofservice)

In [ ]:
vehicles['DISTANCE'].sum()

Interesting observation here. I'm not sure if talked about before but it's intersting to see the total distance is so high. The documentation states that it either represents km/miles but the number is too high when added together. But if you treat the total sum as metres then convert it to km then it will be around 48 million km which is close to the 2019 data Brian mentioned in the presentation. Worth discussing over I feel.

# Data Quality Report 

### Trips

**Feature  - &emsp;Issue &emsp; - Solution**
<br>
* DATASOURCE - constant column - remove feature
<br>
* DAYOFSERVICE - 5 days from 365 days - investigate if the missing 5 days are holidays
<br>
* TRIPID -  None detected - no action needed
<br>
* LINEID - None detected - no action needed
<br>
* ROUTEID - None detected - no action needed
<br>
* DIRECTION - None detected - no action needed
<br>
* PLANNEDTIME_DEP - None detected - no action needed
<br>
* PLANNEDTIME_ARR - None detected - no action needed
<br>
* ACTUALTIME_DEP - 6% missing values - investigate and find average and impute
<br>
* ACTUALTIME_ARR - 7% missing values - investigate and find average and impute
<br>
* BASIN - constant column - remove feature
<br>
* TENDERLOT - 100% missing values - investigate and remove feature
<br>
* SUPPRESSED - 99% missing values - investigate and remove feature
<br>
* JUSTIFICATIONID - 99% missing values - investigate and remove feature
<br>
* LASTUPDATE - None detected - no action needed
<br>
* NOTE - None detected - investigate
<br>
* Trips from GTFS feed - nonexisting routes on old timetable - insert new features and finalize dataset
<br>

# Data Quality Plan 

### Trips

**DATASOURCE:** The feature only has one unique value which is DB. Since we know that the source is constant throughout, we will remove this feature from the dataframe.

In [ ]:
trips.pop('DATASOURCE')

In [ ]:
trips.head(5)

**DAYOFSERVICE:** We see that there are 5 days that aren't accounted for. Let us see what dates that weren't included. 

The dates that were not included:
* March 1 2018
* March 2 2018
* December 9 2018
* December 10 2018
* December 25 2018

This just means that there were no trips planned for this day. Will it affect the prediction model? Probably not. 

In [ ]:
daysofservice = trips['DAYOFSERVICE']
daysofservice = daysofservice.sort_values()

In [ ]:
#daysofservice.unique()

In [ ]:
pd.date_range(start = '2018-01-01', end = '2018-12-31' ).difference(daysofservice)

**ACTUALTIME_DEP:** We will investigate the 6% values missing. First, we will make a dataframe that only have the rows that have ACTUALTIME_DEP missing.

It is important that these missing values are dealt with. In order to do that, leavetimes table will be loaded. I'm going to try and load an SQL query. The query is: "For every TRIPID, load the maximum PROGRNUMBER and return the rows"

We see here at the first 60 rows it is unusual that we only see PROGRNUMBER. It looks as if might be a mistake. However after investigating there are some trips that only had one stop and there were no further records. 

Then, we will make a seperate dataframe that will only have rows of the tripid's from leavetimes table. Then, replace trips' actualtime_dep with leavetimes' actualtime_dep. The dataframe should match 

In [ ]:
actual_dep = trips[trips['ACTUALTIME_DEP'].isna()]
tripday = [[day for day in actual_dep.DAYOFSERVICE], [id_ for id_ in actual_dep.TRIPID]]

In [ ]:
# Converting to timedelta
leavetimes['PLANNEDTIME_ARR'] = leavetimes['PLANNEDTIME_ARR'].apply(pd.to_timedelta, unit='s')
leavetimes['PLANNEDTIME_ARR'] = leavetimes['PLANNEDTIME_ARR']- pd.to_timedelta(leavetimes['PLANNEDTIME_ARR'].dt.days, unit='d')
leavetimes['PLANNEDTIME_DEP'] = leavetimes['PLANNEDTIME_DEP'].apply(pd.to_timedelta, unit='s')
leavetimes['PLANNEDTIME_DEP'] = leavetimes['PLANNEDTIME_DEP']- pd.to_timedelta(leavetimes['PLANNEDTIME_DEP'].dt.days, unit='d')
leavetimes['ACTUALTIME_DEP'] = leavetimes['ACTUALTIME_DEP'].apply(pd.to_timedelta, unit='s')
leavetimes['ACTUALTIME_DEP'] = leavetimes['ACTUALTIME_DEP']- pd.to_timedelta(leavetimes['ACTUALTIME_DEP'].dt.days, unit='d')

In [ ]:
leavetimes.shape

In [ ]:
# Making an empty dataframe
leavetrips = pd.DataFrame()
# df.append(leavetimes[leavetimes['TRIPID'].isin(tripday[1])])

In [ ]:
leave_times = leave_times[(leave_times['TRIPID'].isin(tripday[1]))]
leave_times = leave_times.sort_values(by='PROGRNUMBER')
leave = leave_times.groupby(['TRIPID'])
leave = leave.tail(1)
leavetrips.append(leave)

In [ ]:
actual_dep = actual_dep.sort_values(by='TRIPID')
leavetrips = leavetrips.sort_values(by='TRIPID')

In [ ]:
actual_dep.head(5)

for val in range(len(tripday[0])):
    row = leavetimes.loc[(leavetimes['DAYOFSERVICE']==tripday[0][val]) & (leavetimes['TRIPID']==tripday[1][val])].sort_values(by='PROGRNUMBER').iloc[[-1]]
    df = df.append(row)

leavetimes.loc[(leavetimes['DAYOFSERVICE']==tripday[0][0]) & (leavetimes['TRIPID']==tripday[1][0])].sort_values(by='PROGRNUMBER').iloc[[-1]]

**ACTUALTIME_ARR - TO DO**

**BASIN:** This was considered a constant column and so this feature will be dropped. After investigating it from data understanding, having a constant column would not affect the prediction model as it only has one value throughout. Therefore, it will be dropped.

In [ ]:
trips.pop('BASIN')

In [ ]:
trips.head(1)

**TENDERLOT:** Since 100% of TENDERLOT values are missing it will most likely be dropped. From the documentation, trying to relate tender lot to business terms. This may refer to whether or not the trip is paid for a service. Makes sense that it's hidden for privacy if this is the case. It doesn't relate much to the project.

In [ ]:
trips.pop('TENDERLOT')
trips.head(1)

**SUPPRESSED:** 99% of SUPPRESSED values are missing. It doesn't look like it gives any meaningful data. Not sure why it has to be suppressed. From the documentation, it looks like the NaN values just means it's partially suppressed which may explain TENDERLOT feature being 100% missing because of personal information. Like TENDERLOT, it doesn't relate much to the project and so, it will be dropped

In [ ]:
trips.pop('SUPPRESSED')
trips.head(1)

**JUSTIFICATIONID:** From the documentation, it looks as if the JUSTIFICATIONID is a primary key that is used to link to another table that we are not given, that represents some kind of fault code. Since the majority of the value is missing, the relevancy of the feature is probably not needed for the model. 

In [ ]:
trips.pop('JUSTIFICATIONID')
trips.head(1)

**NOTE:** I will investigate the meaning of the numbers

In [ ]:
trips.dtypes

In [ ]:
# Convert NOTE to type int64
trips['NOTE'].nunique()